compiler/optimizer/OMRLocalCSE.cpp

```diff
@@ -777,8 +777,11 @@ void OMR::LocalCSE::doCommoningIfAvailable(TR::Node *node, TR::Node *parent, int
        shouldCommonNode(parent, node) &&
        performTransformation(comp(), "%s   Local Common Subexpression Elimination commoning node : %p by available node : %p\n", optDetailString(), node, availableExpression))
       {
+      // We can allow final non volatile fields to be commoned even during the volatile only phase
+      // as we do not expect those field to be changing. This enables up to common volatiles that are based on an 
+      // indirection chain of such final non volatiles (or autos or parms that are not global by definition)
```

Comment:
So I'm worried that this is not generally safe as described in the comment. The comment says we can allow final non volatile fields to be commoned during the volatile only phase. I do not believe this is correct. The volatile field read and writes create happens before relationships in the memory model - that is you must re-read fields when crossing a volatile memory operation. If we allow the commoning of non-volatile fields to happen in the volatile only phase we may get this wrong.

Autos and Parms should be fine.

Reply:
I just noticed, I missed to add `final` word before `non volatile` fields, The change does check if the non-volatile is final or auto/parms. Let me fix the comment.

Reply:
We are applying this transformation to hot and above level, where a volatile load based on indirect load chain of final/auto/parm can be commonned up during the volatile only phase. And after reading through the comment in same opt (https://github.com/eclipse/omr/blob/fd82a51b1eeb3903f72dcbbfbb84ecfeb6d0fa36/compiler/optimizer/OMRLocalCSE.cpp#L1197-L1203) I added this comment here.

If the field is Final, Can't we expect that normal reading of this field won't change?

Reply:
Right but the deduplication/commoning we are doing here changes the happens before points so allowing even finals to common outside of the volatile commoning itself could be problematic since commoning the volatile on the dependent chain will delay the change being observed, but the memory model allows that. Commoning a final field read across a volatile read that does not end up being removed would allow a memory ordering violation to occur.

Reply:
This comment needs to be updated I think

Reply:
removed comment in https://github.com/eclipse/omr/compare/10527798af1e29f001de3b99d9ef0e3d1240de19..4c9c4fca399f6bdef3baca5dd21477cf847d30bd and moved the updated comment to the query which to inspect if a non-volatile node can be commoned in volatile pass or not.